File: src/ledger.py
import hashlib
from datetime import datetime, timezone
from typing import List, Optional
from src.schemas import AuditLedgerEntry, RecoveryState
from src.constants import LEDGER_GENESIS_HASH
from src.utils import utc_now_iso

class AuditLedger:
    def __init__(self):
        self.chain: List[AuditLedgerEntry] = []
        self.prev_hash: str = LEDGER_GENESIS_HASH

    def compute_hash(self, entity_id: str, status: str, recovered_paise: int, prev_hash: str, timestamp: str = "") -> str:
        payload = f"{entity_id}:{status}:{recovered_paise}:{prev_hash}:{timestamp}"
        return hashlib.sha256(payload.encode()).hexdigest()

    def record_entry(
        self,
        entity_id: str,
        initial_paise: int,
        recovered_paise: int,
        status: RecoveryState,
        attempt_count: int,
        cost_paise: int,
        reason_code: Optional[str] = "POLICY_EXECUTION",
    ) -> AuditLedgerEntry:
        if initial_paise < 0:
            raise ValueError("initial_paise must be non-negative")
        if recovered_paise < 0:
            raise ValueError("recovered_paise must be non-negative")
        if cost_paise < 0:
            raise ValueError("cost_paise must be non-negative")
        ts = utc_now_iso()
        current_hash = self.compute_hash(entity_id, status.value, recovered_paise, self.prev_hash, ts)
        entry = AuditLedgerEntry(
            log_id=f"log_{len(self.chain) + 1:04d}",
            timestamp=ts,
            entity_id=entity_id,
            initial_amount_paise=initial_paise,
            recovered_amount_paise=recovered_paise,
            status=status,
            attempt_count=attempt_count,
            total_cost_incurred_paise=cost_paise,
            audit_hash=current_hash,
            reason_code=reason_code or "POLICY_EXECUTION",
        )
        self.chain.append(entry)
        self.prev_hash = current_hash
        return entry

    def verify_integrity(self) -> bool:
        prev = LEDGER_GENESIS_HASH
        for entry in self.chain:
            expected_hash = self.compute_hash(
                entry.entity_id,
                entry.status.value,
                entry.recovered_amount_paise,
                prev,
                entry.timestamp,
            )
            if entry.audit_hash != expected_hash:
                return False
            prev = entry.audit_hash
        return True

    def get_summary(self) -> dict:
        total_exposed = 0
        total_recovered = 0
        total_cost = 0
        recovered_count = 0

        for e in self.chain:
            total_exposed += e.initial_amount_paise
            total_recovered += e.recovered_amount_paise
            total_cost += e.total_cost_incurred_paise
            if e.recovered_amount_paise > 0:
                recovered_count += 1

        yield_rate = (total_recovered / total_exposed * 100) if total_exposed > 0 else 0.0
        return {
            "total_records": len(self.chain),
            "total_exposed_gmv_paise": total_exposed,
            "total_recovered_gmv_paise": total_recovered,
            "yield_rate_percent": round(yield_rate, 2),
            "total_cost_paise": total_cost,
            "recovered_count": recovered_count,
            "integrity_valid": self.verify_integrity(),
        }

    def verify_block_proof(self, log_id: str) -> Optional[dict]:
        prev = LEDGER_GENESIS_HASH
        for entry in self.chain:
            expected_hash = self.compute_hash(
                entry.entity_id,
                entry.status.value,
                entry.recovered_amount_paise,
                prev,
                entry.timestamp,
            )
            if entry.log_id == log_id:
                return {
                    "log_id": entry.log_id,
                    "entity_id": entry.entity_id,
                    "status": entry.status.value,
                    "recovered_paise": entry.recovered_amount_paise,
                    "prev_hash": prev,
                    "audit_hash": entry.audit_hash,
                    "recomputed_hash": expected_hash,
                    "is_valid": entry.audit_hash == expected_hash,
                    "reason_code": entry.reason_code,
                }
            prev = entry.audit_hash
        return None

File: src/engine/ledger.ts
import { AuditLedgerEntry, RecoveryState } from './types';
import { LEDGER_GENESIS_HASH } from './constants';
import { sha256Sync } from './utils';

export class AuditLedger {
  public chain: AuditLedgerEntry[] = [];

  public computeHash(entry: Partial<AuditLedgerEntry>, prevHash: string): string {
    const payload = `${entry.log_id}:${entry.timestamp}:${entry.entity_id}:${entry.initial_amount_paise}:${entry.recovered_amount_paise}:${entry.status}:${entry.attempt_count}:${entry.total_cost_incurred_paise}:${entry.reason_code}:${prevHash}`;
    return sha256Sync(payload);
  }

  public recordEntry(
    entity_id: string,
    initial_amount_paise: number,
    recovered_amount_paise: number,
    status: RecoveryState,
    attempt_count: number,
    total_cost_incurred_paise: number,
    reason_code: string = "RECOVERY_ATTEMPT"
  ): AuditLedgerEntry {
    const prevHash = this.chain.length > 0 ? this.chain[this.chain.length - 1].audit_hash : LEDGER_GENESIS_HASH;
    const log_id = `block_${String(this.chain.length + 1).padStart(5, '0')}`;
    const timestamp = new Date().toISOString();

    const partial = {
      log_id,
      timestamp,
      entity_id,
      initial_amount_paise,
      recovered_amount_paise,
      status,
      attempt_count,
      total_cost_incurred_paise,
      reason_code,
    };

    const audit_hash = this.computeHash(partial, prevHash);

    const entry: AuditLedgerEntry = {
      ...partial,
      audit_hash,
    };

    this.chain.push(entry);
    return entry;
  }

  public verifyIntegrity(): boolean {
    if (this.chain.length === 0) return true;

    for (let i = 0; i < this.chain.length; i++) {
      const current = this.chain[i];
      const prevHash = i === 0 ? LEDGER_GENESIS_HASH : this.chain[i - 1].audit_hash;
      const expectedHash = this.computeHash(current, prevHash);
      if (current.audit_hash !== expectedHash) {
        return false;
      }
    }
    return true;
  }

  public verifyBlockProof(log_id: string): { is_valid: boolean; proof_details: any } {
    const idx = this.chain.findIndex(b => b.log_id === log_id);
    if (idx === -1) {
      return { is_valid: false, proof_details: { error: "Block not found" } };
    }

    const block = this.chain[idx];
    const prevHash = idx === 0 ? LEDGER_GENESIS_HASH : this.chain[idx - 1].audit_hash;
    const expectedHash = this.computeHash(block, prevHash);
    const isValid = block.audit_hash === expectedHash;

    return {
      is_valid: isValid,
      proof_details: {
        log_id: block.log_id,
        block_height: idx + 1,
        audit_hash: block.audit_hash,
        previous_block_hash: prevHash,
        recomputed_hash: expectedHash,
        entity_id: block.entity_id,
        initial_amount_paise: block.initial_amount_paise,
        recovered_amount_paise: block.recovered_amount_paise,
        status: block.status,
        cryptographic_proof_status: isValid ? "UNBROKEN_CRYPTOGRAPHIC_INTEGRITY" : "HASH_MISMATCH",
      }
    };
  }

  public getSummary(): {
    total_records: number;
    total_initial_paise: number;
    total_recovered_paise: number;
    total_cost_paise: number;
    net_recovered_paise: number;
    recovery_rate_pct: number;
    integrity_valid: boolean;
  } {
    let totalInitial = 0;
    let totalRecovered = 0;
    let totalCost = 0;

    for (const e of this.chain) {
      totalInitial += e.initial_amount_paise;
      totalRecovered += e.recovered_amount_paise;
      totalCost += e.total_cost_incurred_paise;
    }

    const netRecovered = totalRecovered - totalCost;
    const recoveryRate = totalInitial > 0 ? (totalRecovered / totalInitial) * 100 : 0;

    return {
      total_records: this.chain.length,
      total_initial_paise: totalInitial,
      total_recovered_paise: totalRecovered,
      total_cost_paise: totalCost,
      net_recovered_paise: netRecovered,
      recovery_rate_pct: Number(recoveryRate.toFixed(2)),
      integrity_valid: this.verifyIntegrity(),
    };
  }

  public clear() {
    this.chain = [];
  }
}
